flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestBucketPartitioner.java

```diff
@@ -0,0 +1,99 @@
+import org.junit.jupiter.params.provider.EnumSource;
+import org.junit.jupiter.params.provider.ValueSource;
+
+public class TestBucketPartitioner {
+
+  static final int NUM_BUCKETS = 60;
+
+  @ParameterizedTest
+  @EnumSource(
+      value = TestBucketPartitionerUtils.TableSchemaType.class,
+      names = {"ONE_BUCKET", "IDENTITY_AND_BUCKET"})
+  public void testPartitioningParallelismGreaterThanBuckets(
+      TestBucketPartitionerUtils.TableSchemaType tableSchemaType) {
+    final int numPartitions = 500;
+
+    PartitionSpec partitionSpec =
+        TestBucketPartitionerUtils.getPartitionSpec(tableSchemaType, NUM_BUCKETS);
+
+    BucketPartitioner bucketPartitioner = new BucketPartitioner(partitionSpec);
+
+    for (int expectedIdx = 0, bucketId = 0; expectedIdx < numPartitions; expectedIdx++) {
+      int actualIdx = bucketPartitioner.partition(bucketId, numPartitions);
+      Assertions.assertThat(actualIdx).isEqualTo(expectedIdx);
+      if (++bucketId == NUM_BUCKETS) {
+        bucketId = 0;
+      }
+    }
+  }
+
+  @ParameterizedTest
+  @EnumSource(
+      value = TestBucketPartitionerUtils.TableSchemaType.class,
+      names = {"ONE_BUCKET", "IDENTITY_AND_BUCKET"})
+  public void testPartitioningParallelismEqualLessThanBuckets(
+      TestBucketPartitionerUtils.TableSchemaType tableSchemaType) {
+    final int numPartitions = 30;
+
+    PartitionSpec partitionSpec =
+        TestBucketPartitionerUtils.getPartitionSpec(tableSchemaType, NUM_BUCKETS);
+
+    BucketPartitioner bucketPartitioner = new BucketPartitioner(partitionSpec);
+
+    for (int bucketId = 0; bucketId < NUM_BUCKETS; bucketId++) {
+      int actualIdx = bucketPartitioner.partition(bucketId, numPartitions);
+      Assertions.assertThat(actualIdx).isEqualTo(bucketId % numPartitions);
+    }
+  }
+
+  @ParameterizedTest
+  @EnumSource(value = TestBucketPartitionerUtils.TableSchemaType.class, names = "TWO_BUCKETS")
+  public void testPartitionerMultipleBucketsFail(
+      TestBucketPartitionerUtils.TableSchemaType tableSchemaType) {
+    PartitionSpec partitionSpec =
+        TestBucketPartitionerUtils.getPartitionSpec(tableSchemaType, NUM_BUCKETS);
+
+    Assertions.assertThatExceptionOfType(RuntimeException.class)
+        .isThrownBy(() -> new BucketPartitioner(partitionSpec))
+        .withMessageContaining(BucketPartitionerUtils.BAD_NUMBER_OF_BUCKETS_ERROR_MESSAGE);
+  }
+
+  @ParameterizedTest
+  @ValueSource(ints = {-1, NUM_BUCKETS})
+  public void testPartitionerBucketIdOutOfRangeFail(int bucketId) {
+    PartitionSpec partitionSpec =
+        TestBucketPartitionerUtils.getPartitionSpec(
+            TestBucketPartitionerUtils.TableSchemaType.ONE_BUCKET, NUM_BUCKETS);
+
+    BucketPartitioner bucketPartitioner = new BucketPartitioner(partitionSpec);
+
+    Assertions.assertThatExceptionOfType(IllegalArgumentException.class)
+        .isThrownBy(() -> bucketPartitioner.partition(bucketId, 1))
+        .withMessageContaining(BUCKET_OUT_OF_RANGE_MESSAGE_PREFIX);
```

Comment:
nit: I haven't seen Iceberg code use constant string for unit test error message. maybe also assert the exact error msg.

Reply:
I actually changed all assertions to check for complete messages